File: Chagok/frontend/src/lib/api/messages.ts
/**
 * Message API client
 * 003-role-based-ui Feature - US6
 */

import { apiClient, ApiResponse } from './client';
import { logger } from '@/lib/logger';
import type {
  Message,
  MessageListResponse,
  ConversationListResponse,
  UnreadCountResponse,
  SendMessageRequest,
  MarkReadRequest,
  DirectMessage,
  DirectMessageSummary,
  DirectMessageListResponse,
  DirectMessageQueryParams,
  DirectMessageCreate,
} from '@/types/message';

const BASE_PATH = '/messages';

interface WebSocketTokenResponse {
  token: string;
  expires_in: number;
}

/**
 * Get list of conversations for the current user
 */
export async function getConversations(): Promise<ApiResponse<ConversationListResponse>> {
  return apiClient.get<ConversationListResponse>(`${BASE_PATH}/conversations`);
}

/**
 * Get unread message count
 */
export async function getUnreadCount(): Promise<ApiResponse<UnreadCountResponse>> {
  return apiClient.get<UnreadCountResponse>(`${BASE_PATH}/unread`);
}

/**
 * Get messages for a case
 */
export async function getMessages(
  caseId: string,
  options?: {
    otherUserId?: string;
    limit?: number;
    beforeId?: string;
  }
): Promise<ApiResponse<MessageListResponse>> {
  const params = new URLSearchParams();
  if (options?.otherUserId) params.set('other_user_id', options.otherUserId);
  if (options?.limit) params.set('limit', String(options.limit));
  if (options?.beforeId) params.set('before_id', options.beforeId);

  const query = params.toString();
  const url = `${BASE_PATH}/${caseId}${query ? `?${query}` : ''}`;

  return apiClient.get<MessageListResponse>(url);
}

/**
 * Send a new message
 */
export async function sendMessage(
  data: SendMessageRequest
): Promise<ApiResponse<Message>> {
  return apiClient.post<Message>(BASE_PATH, data);
}

/**
 * Mark messages as read
 */
export async function markMessagesRead(
  messageIds: string[]
): Promise<ApiResponse<{ marked_count: number }>> {
  const body: MarkReadRequest = { message_ids: messageIds };
  return apiClient.post<{ marked_count: number }>(`${BASE_PATH}/read`, body);
}

/**
 * Issue a short-lived token for WebSocket authentication
 */
export async function getWebSocketToken(): Promise<ApiResponse<WebSocketTokenResponse>> {
  return apiClient.get<WebSocketTokenResponse>(`${BASE_PATH}/ws-token`);
}

// ============== WebSocket Client ==============

export type WebSocketStatus = 'connecting' | 'connected' | 'disconnected' | 'error';

export interface MessageWebSocketCallbacks {
  onMessage?: (message: Message) => void;
  onTyping?: (data: { user_id: string; case_id: string; is_typing: boolean }) => void;
  onReadReceipt?: (messageIds: string[]) => void;
  onOfflineMessages?: (messages: Message[]) => void;
  onStatusChange?: (status: WebSocketStatus) => void;
  onError?: (error: string) => void;
}

/**
 * Create a WebSocket connection for real-time messaging
 */
const buildSocketUrl = (token: string): string => {
  const apiBaseUrl = process.env.NEXT_PUBLIC_API_BASE_URL || 'http://localhost:8000';
  const wsProtocol = apiBaseUrl.startsWith('https') ? 'wss' : 'ws';
  const wsHost = apiBaseUrl.replace(/^https?:\/\//, '');
  return `${wsProtocol}://${wsHost}${BASE_PATH}/ws?token=${encodeURIComponent(token)}`;
};

export function buildWebSocketUrl(token: string): string {
  return buildSocketUrl(token);
}

export function createMessageWebSocket(
  token: string,
  callbacks: MessageWebSocketCallbacks
): {
  send: (type: string, payload: unknown) => void;
  sendMessage: (caseId: string, recipientId: string, content: string, attachments?: string[]) => void;
  sendTyping: (caseId: string, recipientId: string, isTyping: boolean) => void;
  markRead: (messageIds: string[]) => void;
  close: () => void;
} {
  const wsUrl = buildSocketUrl(token);
  let ws: WebSocket | null = null;
  let reconnectAttempts = 0;
  const maxReconnectAttempts = 5;
  let reconnectTimeout: NodeJS.Timeout | null = null;

  const connect = () => {
    callbacks.onStatusChange?.('connecting');

    ws = new WebSocket(wsUrl);

    ws.onopen = () => {
      reconnectAttempts = 0;
      callbacks.onStatusChange?.('connected');
    };

    ws.onclose = (event) => {
      callbacks.onStatusChange?.('disconnected');

      // Attempt to reconnect if not a normal closure
      if (event.code !== 1000 && reconnectAttempts < maxReconnectAttempts) {
        reconnectAttempts++;
        const delay = Math.min(1000 * Math.pow(2, reconnectAttempts), 30000);
        reconnectTimeout = setTimeout(connect, delay);
      }
    };

    ws.onerror = () => {
      callbacks.onStatusChange?.('error');
      callbacks.onError?.('WebSocket connection error');
    };

    ws.onmessage = (event) => {
      try {
        const data = JSON.parse(event.data);
        const { type, payload } = data;

        switch (type) {
          case 'new_message':
          case 'message_sent':
            callbacks.onMessage?.(payload as Message);
            break;

          case 'typing':
            callbacks.onTyping?.(payload);
            break;

          case 'read_receipt':
          case 'read_confirmed':
            callbacks.onReadReceipt?.(payload.message_ids);
            break;

          case 'offline_messages':
            callbacks.onOfflineMessages?.(payload.messages);
            break;

          case 'error':
            callbacks.onError?.(payload.message);
            break;

          case 'pong':
            // Keep-alive response, no action needed
            break;
        }
      } catch {
        logger.error('Failed to parse WebSocket message');
      }
    };
  };

  connect();

  const send = (type: string, payload: unknown) => {
    if (ws?.readyState === WebSocket.OPEN) {
      ws.send(JSON.stringify({ type, payload }));
    }
  };

  return {
    send,

    sendMessage: (caseId: string, recipientId: string, content: string, attachments?: string[]) => {
      send('message', {
        case_id: caseId,
        recipient_id: recipientId,
        content,
        attachments,
      });
    },

    sendTyping: (caseId: string, recipientId: string, isTyping: boolean) => {
      send('typing', {
        case_id: caseId,
        recipient_id: recipientId,
        is_typing: isTyping,
      });
    },

    markRead: (messageIds: string[]) => {
      send('read', { message_ids: messageIds });
    },

    close: () => {
      if (reconnectTimeout) {
        clearTimeout(reconnectTimeout);
      }
      if (ws) {
        ws.close(1000, 'User closed connection');
        ws = null;
      }
    },
  };
}

// ============== Direct Message API (US2 - Lawyer Portal) ==============

const DIRECT_MESSAGE_PATH = '/direct-messages';

/**
 * Get list of direct messages (inbox or sent)
 */
export async function getDirectMessages(
  params?: DirectMessageQueryParams
): Promise<ApiResponse<DirectMessageListResponse>> {
  const searchParams = new URLSearchParams();

  if (params?.folder) {
    searchParams.append('folder', params.folder);
  }
  if (params?.page) {
    searchParams.append('page', String(params.page));
  }
  if (params?.limit) {
    searchParams.append('limit', String(params.limit));
  }

  const queryString = searchParams.toString();
  const url = queryString ? `${DIRECT_MESSAGE_PATH}?${queryString}` : DIRECT_MESSAGE_PATH;

  return apiClient.get<DirectMessageListResponse>(url);
}

/**
 * Get a single direct message by ID
 */
export async function getDirectMessage(
  messageId: string
): Promise<ApiResponse<DirectMessage>> {
  return apiClient.get<DirectMessage>(`${DIRECT_MESSAGE_PATH}/${messageId}`);
}

/**
 * Send a new direct message
 */
export async function sendDirectMessage(
  data: DirectMessageCreate
): Promise<ApiResponse<DirectMessage>> {
  return apiClient.post<DirectMessage>(DIRECT_MESSAGE_PATH, data);
}

/**
 * Delete a direct message (soft delete for current user)
 */
export async function deleteDirectMessage(
  messageId: string
): Promise<ApiResponse<{ success: boolean }>> {
  return apiClient.delete<{ success: boolean }>(`${DIRECT_MESSAGE_PATH}/${messageId}`);
}

/**
 * Mark a direct message as read
 */
export async function markDirectMessageAsRead(
  messageId: string
): Promise<ApiResponse<DirectMessage>> {
  return apiClient.patch<DirectMessage>(`${DIRECT_MESSAGE_PATH}/${messageId}/read`, {});
}

// Re-export direct message types
export type {
  DirectMessage,
  DirectMessageSummary,
  DirectMessageListResponse,
  DirectMessageQueryParams,
  DirectMessageCreate,
};

File: Chagok/backend/app/api/messages.py
"""
Message API endpoints with WebSocket support
003-role-based-ui Feature - US6

REST Endpoints:
- GET /messages/conversations - List conversations
- GET /messages/{case_id} - Get messages for a case
- POST /messages - Send a message
- POST /messages/read - Mark messages as read
- GET /messages/unread - Get unread count

WebSocket:
- WS /messages/ws - Real-time message channel
"""

import asyncio
import logging
from datetime import datetime, timedelta
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, WebSocket, WebSocketDisconnect, Query
from sqlalchemy.orm import Session

from app.db.session import get_db
from app.db.models import User
from app.core.dependencies import get_current_user, get_current_user_id
from app.core.security import decode_access_token, create_access_token
from app.services.message_service import MessageService
from app.schemas.message import (
    MessageCreate,
    MessageMarkRead,
    MessageResponse,
    MessageListResponse,
    ConversationListResponse,
    UnreadCountResponse,
)
from app.db.schemas import (
    MessageCreate as MessageCreateV2,
    MessageResponse as MessageResponseV2,
    MessageListResponse as MessageListResponseV2,
)

logger = logging.getLogger(__name__)

router = APIRouter()


# ============== Connection Manager for WebSocket ==============

class ConnectionManager:
    """
    Manages WebSocket connections for real-time messaging.

    Features:
    - User connection tracking
    - Broadcast to specific users
    - Presence updates
    """

    def __init__(self):
        # user_id -> list of WebSocket connections (user can have multiple tabs)
        self.active_connections: dict[str, list[WebSocket]] = {}
        # user_id -> last seen timestamp
        self.presence: dict[str, datetime] = {}

    async def connect(self, websocket: WebSocket, user_id: str):
        """Accept connection and register user."""
        await websocket.accept()
        if user_id not in self.active_connections:
            self.active_connections[user_id] = []
        self.active_connections[user_id].append(websocket)
        self.presence[user_id] = datetime.utcnow()

    def disconnect(self, websocket: WebSocket, user_id: str):
        """Remove connection on disconnect."""
        if user_id in self.active_connections:
            if websocket in self.active_connections[user_id]:
                self.active_connections[user_id].remove(websocket)
            if not self.active_connections[user_id]:
                del self.active_connections[user_id]
                del self.presence[user_id]

    async def send_to_user(self, user_id: str, message: dict):
        """Send message to all connections of a specific user."""
        if user_id in self.active_connections:
            for connection in self.active_connections[user_id]:
                try:
                    await connection.send_json(message)
                except Exception:
                    # Connection might be stale, will be cleaned up on next disconnect
                    pass

    async def broadcast_to_users(self, user_ids: list[str], message: dict):
        """Send message to multiple users."""
        for user_id in user_ids:
            await self.send_to_user(user_id, message)

    def is_online(self, user_id: str) -> bool:
        """Check if user is currently online."""
        return user_id in self.active_connections

    def get_online_users(self) -> list[str]:
        """Get list of online user IDs."""
        return list(self.active_connections.keys())


# Global connection manager instance
manager = ConnectionManager()


# ============== REST Endpoints ==============

@router.get("/conversations", response_model=ConversationListResponse)
def get_conversations(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """
    Get list of conversations for the current user.

    Returns conversations grouped by case and other user,
    with last message preview and unread count.
    """
    service = MessageService(db)
    return service.get_conversations(current_user.id)


@router.get("/unread", response_model=UnreadCountResponse)
def get_unread_count(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """
    Get unread message count.

    Returns total unread and breakdown by case.
    """
    service = MessageService(db)
    return service.get_unread_count(current_user.id)


@router.get("/{case_id}", response_model=MessageListResponse)
def get_messages(
    case_id: str,
    other_user_id: Optional[str] = Query(None, description="Filter to conversation with this user"),
    limit: int = Query(50, ge=1, le=100, description="Maximum messages to return"),
    before_id: Optional[str] = Query(None, description="Get messages before this ID"),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """
    Get messages for a case.

    Optionally filter to a specific conversation with another user.
    Supports pagination with before_id.
    """
    service = MessageService(db)
    try:
        return service.get_messages(
            user_id=current_user.id,
            case_id=case_id,
            other_user_id=other_user_id,
            limit=limit,
            before_id=before_id,
        )
    except PermissionError as e:
        logger.warning(f"Permission denied for user {current_user.id} accessing messages: {e}")
        raise HTTPException(status_code=403, detail="메시지에 접근할 권한이 없습니다")


@router.post("", response_model=MessageResponse)
async def send_message(
    message_data: MessageCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """
    Send a new message.

    Also broadcasts to recipient via WebSocket if online.
    """
    service = MessageService(db)
    try:
        message = service.send_message(
            sender_id=current_user.id,
            message_data=message_data,
        )

        # Broadcast via WebSocket to recipient
        await manager.send_to_user(
            message_data.recipient_id,
            {
                "type": "new_message",
                "payload": message.model_dump(mode="json"),
            },
        )

        return message
    except PermissionError as e:
        logger.warning(f"Permission denied for user {current_user.id} sending message: {e}")
        raise HTTPException(status_code=403, detail="메시지 전송 권한이 없습니다")
    except ValueError as e:
        logger.warning(f"Validation error sending message: {e}")
        raise HTTPException(status_code=400, detail="메시지 전송에 실패했습니다. 입력값을 확인해주세요")


@router.post("/read")
def mark_messages_read(
    data: MessageMarkRead,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """
    Mark messages as read.

    Only marks messages where the current user is the recipient.
    """
    service = MessageService(db)
    count = service.mark_as_read(current_user.id, data.message_ids)
    return {"marked_count": count}


@router.get("/ws-token")
def get_ws_token(
    current_user: User = Depends(get_current_user),
):
    """
    Issue a short-lived token for WebSocket authentication.

    Returns:
        token: JWT string scoped for WebSocket connections
        expires_in: Expiration time in seconds
    """
    token = create_access_token(
        {
            "sub": current_user.id,
            "role": current_user.role.value,
            "token_type": "ws",
        },
        expires_delta=timedelta(minutes=5)
    )
    return {"token": token, "expires_in": 5 * 60}


# ============== WebSocket Endpoint ==============

@router.websocket("/ws")
async def websocket_endpoint(
    websocket: WebSocket,
    token: Optional[str] = Query(None),
    db: Session = Depends(get_db),
):
    """
    WebSocket endpoint for real-time messaging.

    Connection:
        ws://host/api/messages/ws?token=<jwt_token>

    Message Types (client -> server):
        - message: Send a new message
        - read: Mark messages as read
        - typing: Typing indicator
        - ping: Keep-alive

    Message Types (server -> client):
        - new_message: New message received
        - read_receipt: Message was read
        - typing: Someone is typing
        - presence: User online/offline
        - offline_messages: Unread messages on connect
        - pong: Keep-alive response
    """
    # Authenticate via token
    if not token:
        await websocket.close(code=4001, reason="Token required")
        return

    payload = decode_access_token(token)
    if not payload:
        await websocket.close(code=4001, reason="Invalid token")
        return

    user_id = payload.get("sub")
    if not user_id:
        await websocket.close(code=4001, reason="Invalid token payload")
        return

    # Connect
    await manager.connect(websocket, user_id)

    try:
        # Send offline messages on connect
        service = MessageService(db)
        offline_messages = service.get_offline_messages(user_id)
        if offline_messages:
            await websocket.send_json({
                "type": "offline_messages",
                "payload": {
                    "messages": [m.model_dump(mode="json") for m in offline_messages],
                },
            })

        # Main message loop
        while True:
            try:
                data = await asyncio.wait_for(
                    websocket.receive_json(),
                    timeout=60.0,  # Timeout for keep-alive
                )
            except asyncio.TimeoutError:
                # Send ping to check connection
                await websocket.send_json({"type": "ping"})
                continue

            msg_type = data.get("type")
            payload_data = data.get("payload", {})

            if msg_type == "message":
                # Handle new message
                try:
                    message_data = MessageCreate(**payload_data)
                    message = service.send_message(user_id, message_data)

                    # Send to recipient
                    await manager.send_to_user(
                        message_data.recipient_id,
                        {
                            "type": "new_message",
                            "payload": message.model_dump(mode="json"),
                        },
                    )

                    # Confirm to sender
                    await websocket.send_json({
                        "type": "message_sent",
                        "payload": message.model_dump(mode="json"),
                    })
                except Exception as e:
                    await websocket.send_json({
                        "type": "error",
                        "payload": {"message": str(e)},
                    })

            elif msg_type == "read":
                # Handle read receipt
                message_ids = payload_data.get("message_ids", [])
                service.mark_as_read(user_id, message_ids)

                # Notify senders
                # (For simplicity, broadcast to all involved - could optimize)
                await websocket.send_json({
                    "type": "read_confirmed",
                    "payload": {"message_ids": message_ids},
                })

            elif msg_type == "typing":
                # Forward typing indicator to recipient
                recipient_id = payload_data.get("recipient_id")
                if recipient_id:
                    await manager.send_to_user(
                        recipient_id,
                        {
                            "type": "typing",
                            "payload": {
                                "user_id": user_id,
                                "case_id": payload_data.get("case_id"),
                                "is_typing": payload_data.get("is_typing", True),
                            },
                        },
                    )

            elif msg_type == "ping":
                await websocket.send_json({"type": "pong"})

    except WebSocketDisconnect:
        manager.disconnect(websocket, user_id)
    except Exception:
        manager.disconnect(websocket, user_id)


# ============================================
# Issue #296 - FR-008: New Message Endpoints
# ============================================
@router.get("/v2/inbox", response_model=MessageListResponseV2)
def get_inbox_v2(
    page: int = Query(1, ge=1, description="Page number"),
    limit: int = Query(20, ge=1, le=100, description="Items per page"),
    db: Session = Depends(get_db),
    user_id: str = Depends(get_current_user_id),
):
    """
    Get received messages (inbox) - Issue #296 FR-008.

    - Returns messages where current user is recipient
    - Excludes soft-deleted messages
    - Ordered by newest first
    """
    service = MessageService(db)
    return service.get_inbox_v2(user_id=user_id, page=page, limit=limit)


@router.get("/v2/sent", response_model=MessageListResponseV2)
def get_sent_v2(
    page: int = Query(1, ge=1, description="Page number"),
    limit: int = Query(20, ge=1, le=100, description="Items per page"),
    db: Session = Depends(get_db),
    user_id: str = Depends(get_current_user_id),
):
    """
    Get sent messages - Issue #296 FR-008.

    - Returns messages where current user is sender
    - Excludes soft-deleted messages
    - Ordered by newest first
    """
    service = MessageService(db)
    return service.get_sent_v2(user_id=user_id, page=page, limit=limit)


@router.post("/v2", response_model=MessageResponseV2, status_code=201)
async def send_message_v2(
    data: MessageCreateV2,
    db: Session = Depends(get_db),
    user_id: str = Depends(get_current_user_id),
):
    """
    Send a new message - Issue #296 FR-008.

    - recipient_id is required
    - subject is optional
    - case_id is optional (can send messages without a case)
    """
    service = MessageService(db)
    try:
        message = service.send_message_v2(sender_id=user_id, data=data)

        # Broadcast via WebSocket to recipient if online
        await manager.send_to_user(
            data.recipient_id,
            {
                "type": "new_message",
                "payload": message.model_dump(mode="json"),
            },
        )

        return message
    except ValueError as e:
        logger.warning(f"Validation error sending message v2: {e}")
        raise HTTPException(status_code=400, detail="메시지 전송에 실패했습니다. 입력값을 확인해주세요")


@router.get("/v2/{message_id}", response_model=MessageResponseV2)
def get_message_v2(
    message_id: str,
    db: Session = Depends(get_db),
    user_id: str = Depends(get_current_user_id),
):
    """
    Get a specific message - Issue #296 FR-008.

    - Returns message details
    - Automatically marks as read if user is recipient
    """
    service = MessageService(db)
    try:
        return service.get_message_v2(message_id=message_id, user_id=user_id)
    except KeyError:
        raise HTTPException(status_code=404, detail="Message not found")
    except PermissionError:
        raise HTTPException(status_code=403, detail="Not authorized to access this message")


@router.delete("/v2/{message_id}", status_code=204)
def delete_message_v2(
    message_id: str,
    db: Session = Depends(get_db),
    user_id: str = Depends(get_current_user_id),
):
    """
    Delete a message (soft delete) - Issue #296 FR-008.

    - Marks as deleted for the current user only
    - Message remains visible for the other party
    """
    service = MessageService(db)
    try:
        service.delete_message_v2(message_id=message_id, user_id=user_id)
    except KeyError:
        raise HTTPException(status_code=404, detail="Message not found")
    except PermissionError:
        raise HTTPException(status_code=403, detail="Not authorized to delete this message")


@router.patch("/v2/{message_id}/read", response_model=MessageResponseV2)
def mark_message_read_v2(
    message_id: str,
    db: Session = Depends(get_db),
    user_id: str = Depends(get_current_user_id),
):
    """
    Mark a message as read - Issue #296 FR-008.

    - Only recipient can mark as read
    """
    service = MessageService(db)
    try:
        return service.mark_as_read_v2(message_id=message_id, user_id=user_id)
    except KeyError:
        raise HTTPException(status_code=404, detail="Message not found")
    except PermissionError:
        raise HTTPException(status_code=403, detail="Only recipient can mark message as read")
